docs/encoding_options.md
---
title: Encodings to tensors
toc-title: Table of contents
---

When getting the data into your model, you usually will have to process
it to some form of structured numeric input.

There is a plethora fo ways to encode information. And I will show a
couple (and how to get them using ms2ml)

## Encoding Peptides

::: {.cell execution_count="1"}
``` {.python .cell-code}
import numpy as np
from ms2ml.peptide import Peptide

p = Peptide.from_sequence("MYPEPTIDE")
print(p)
```

::: {.cell-output .cell-output-stdout}
    Peptide.from_sequence('MYPEPTIDE')
:::

::: {.cell-output .cell-output-stderr}
    /Users/sebastianpaez/Library/Caches/pypoetry/virtualenvs/ms2ml-zJc7qiTs-py3.9/lib/python3.9/site-packages/ms2ml/config.py:392: UserWarning: Using default config. Consider creating your own.
      warnings.warn("Using default config. Consider creating your own.", UserWarning)
:::
:::

### Counting

::: {.cell execution_count="2"}
``` {.python .cell-code}
count = p.aa_to_count()
print(count)
```

::: {.cell-output .cell-output-stdout}
    [1 0 0 0 1 2 0 0 0 1 0 0 0 1 0 0 2 0 0 0 1 0 0 0 0 1 0 1 0]
:::
:::

### One-hot encoding

::: {.cell execution_count="3"}
``` {.python .cell-code}
oh = p.aa_to_onehot()
# Note that there is an equivalent method for the modifications
# p.mod_to_onehot()
print(oh)
```

::: {.cell-output .cell-output-stdout}
    [[1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
     [0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
     [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0]
     [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0]
     [0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
     [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0]
     [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0]
     [0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
     [0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
     [0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
     [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0]]
:::
:::

### Vectorizing

::: {.cell execution_count="4"}
``` {.python .cell-code}
vec = p.aa_to_vector()
print(vec)
```

::: {.cell-output .cell-output-stdout}
    [ 0 13 25 16  5 16 20  9  4  5 27]
:::
:::

## Encoding Spectra

::: {.cell execution_count="5"}
``` {.python .cell-code}
from ms2ml.spectrum import Spectrum
from ms2ml.config import Config

spectrum = Spectrum(
    mz=np.array([1000.0, 1500.0, 2000.0]),
    intensity=np.array([1.0, 2.0, 3.0]),
    ms_level=2,
    precursor_mz=1000.0,
)
print(spectrum)
```

::: {.cell-output .cell-output-stdout}
    Spectrum(mz=array([1000., 1500., 2000.]), intensity=array([1., 2., 3.]), ms_level=2, precursor_mz=1000.0, precursor_charge=None, instrument=None, analyzer=None, collision_energy=nan, activation=None, extras={}, retention_time=RetentionTime(rt=nan, units='seconds', run=None), precursor_ion_mobility=None)
:::
:::

### Binning

::: {.cell execution_count="6"}
``` {.python .cell-code}
binned = spectrum.bin_spectrum(
    start = 500.0,
    end = 1500.0,
    binsize = 100.0, # For the real world this would be 1 or 0.02 ...
)

# or ....

binned = spectrum.bin_spectrum(
    start = 500.0,
    end = 1500.0,
    n_bins = 10,
)
print(binned)
```

::: {.cell-output .cell-output-stdout}
    [0. 0. 0. 0. 1. 0. 0. 0. 2.]
:::
:::

::: {.cell execution_count="7"}
``` {.python .cell-code}
# or ....
binned = spectrum.bin_spectrum(
    start = -500.0,
    end = +500.0,
    binsize = 100.0,
    relative = True, # Will make all bins relative to the precursor mz
)

# or ....
binned = spectrum.bin_spectrum(
    start = -500.0,
    end = +500.0,
    binsize = 100.0,
    relative = 600, # Will make all bins relative to 600!
    get_breaks = True
)
binned_spectrum, bin_breaks = binned
print(binned_spectrum)
print(bin_breaks)
```

::: {.cell-output .cell-output-stdout}
    [0. 0. 1. 0. 0. 0. 0. 2. 0.]
    [ 100.          211.11111111  322.22222222  433.33333333  544.44444444
      655.55555556  766.66666667  877.77777778  988.88888889 1100.        ]
:::
:::

## Encoding annotated spectra

::: {.cell execution_count="8"}
``` {.python .cell-code}
from ms2ml.spectrum import AnnotatedPeptideSpectrum
config = Config()
peptide = Peptide.from_sequence("PEPPINK/2", config)
spectrum = AnnotatedPeptideSpectrum(
    mz=np.array([50.0, 147.11333, 1000.0, 1500.0, 2000.0]),
    intensity=np.array([50.0, 200.0, 1.0, 2.0, 3.0]),
    ms_level=2,
    precursor_peptide=peptide,
    precursor_mz=147.11333,
)
spectrum
print(spectrum)
```

::: {.cell-output .cell-output-stdout}
    AnnotatedPeptideSpectrum(mz=array([  50.     ,  147.11333, 1000.     , 1500.     , 2000.     ]), intensity=array([ 50., 200.,   1.,   2.,   3.]), ms_level=2, precursor_mz=147.11333, precursor_charge=2, instrument=None, analyzer=None, collision_energy=nan, activation=None, extras={}, retention_time=RetentionTime(rt=nan, units='seconds', run=None), precursor_ion_mobility=None, precursor_peptide=Peptide([('P', None), ('E', None), ('P', None), ('P', None), ('I', None), ('N', None), ('K', None)], {'n_term': None, 'c_term': None, 'unlocalized_modifications': [], 'labile_modifications': [], 'fixed_modifications': [], 'intervals': [], 'isotopes': [], 'group_ids': [], 'charge_state': ChargeState(2, [])}), precursor_isotope=0)
:::
:::

::: {.cell execution_count="9"}
``` {.python .cell-code}
spectrum.encode_fragments()
```

::: {.cell-output .cell-output-stderr}
    2023-10-27 15:44:47.839 | DEBUG    | ms2ml.proforma_utils:mod_id_mass:117 - Resolving mod_id: 21
    2023-10-27 15:44:47.839 | DEBUG    | ms2ml.proforma_utils:solver:84 - Initializing UnimodResolver
    2023-10-27 15:44:48.048 | DEBUG    | ms2ml.proforma_utils:mod_id_mass:119 - Resolved to {'date_time_modified': datetime.datetime(2018, 8, 13, 13, 42, 59), 'date_time_posted': datetime.datetime(2002, 8, 19, 19, 17, 11), 'avge_mass': 79.9799, 'mono_mass': 79.966331, 'composition': Composition({'H': 1, 'O': 3, 'P': 1}), 'record_id': 21, 'approved': True, 'title': 'Phospho', 'full_name': 'Phosphorylation', 'username_of_poster': 'unimod', 'group_of_poster': 'admin', 'specificity': [{'hidden': True, 'spec_group': 8, 'site': 'E', 'position': 'Anywhere', 'classification': 'Post-translational'}, {'hidden': True, 'spec_group': 6, 'site': 'R', 'position': 'Anywhere', 'classification': 'Post-translational'}, {'hidden': True, 'spec_group': 7, 'site': 'K', 'position': 'Anywhere', 'classification': 'Post-translational'}, {'hidden': True, 'spec_group': 4, 'site': 'H', 'position': 'Anywhere', 'classification': 'Post-translational', 'note': 'Rare'}, {'hidden': True, 'spec_group': 5, 'site': 'C', 'position': 'Anywhere', 'classification': 'Post-translational', 'note': 'Rare'}, {'hidden': True, 'spec_group': 3, 'site': 'D', 'position': 'Anywhere', 'classification': 'Post-translational', 'note': 'Rare'}, {'hidden': False, 'spec_group': 2, 'site': 'Y', 'position': 'Anywhere', 'classification': 'Post-translational'}, {'hidden': False, 'spec_group': 1, 'site': 'T', 'position': 'Anywhere', 'classification': 'Post-translational'}, {'hidden': False, 'spec_group': 1, 'site': 'S', 'position': 'Anywhere', 'classification': 'Post-translational'}], 'refs': [{'text': 'AA0036', 'source': 'RESID', 'url': None}, {'text': 'IonSource', 'source': 'Misc. URL', 'url': 'http://www.ionsource.com/Card/phos/phos.htm'}, {'text': 'AA0037', 'source': 'RESID', 'url': None}, {'text': 'AA0033', 'source': 'RESID', 'url': None}, {'text': 'AA0038', 'source': 'RESID', 'url': None}, {'text': 'AA0039', 'source': 'RESID', 'url': None}, {'text': 'AA0222', 'source': 'RESID', 'url': None}, {'text': 'PHOS', 'source': 'FindMod', 'url': None}, {'text': 'AA0034', 'source': 'RESID', 'url': None}, {'text': 'AA0035', 'source': 'RESID', 'url': None}]}
    2023-10-27 15:44:48.048 | DEBUG    | ms2ml.proforma_utils:mod_id_mass:117 - Resolving mod_id: 7
    2023-10-27 15:44:48.048 | DEBUG    | ms2ml.proforma_utils:mod_id_mass:119 - Resolved to {'date_time_modified': datetime.datetime(2018, 10, 25, 9, 32, 26), 'date_time_posted': datetime.datetime(2002, 8, 19, 19, 17, 11), 'avge_mass': 0.9848, 'mono_mass': 0.984016, 'composition': Composition({'H': -1, 'N': -1, 'O': 1}), 'record_id': 7, 'approved': False, 'title': 'Deamidated', 'full_name': 'Deamidation', 'username_of_poster': 'unimod', 'group_of_poster': 'admin', 'specificity': [{'hidden': False, 'spec_group': 1, 'site': 'Q', 'position': 'Anywhere', 'classification': 'Artefact'}, {'hidden': True, 'spec_group': 2, 'site': 'R', 'position': 'Anywhere', 'classification': 'Post-translational', 'note': 'Protein which is post-translationally modified by the de-imination of one or more arginine residues; Peptidylarginine deiminase (PAD) converts protein bound to citrulline'}, {'hidden': False, 'spec_group': 1, 'site': 'N', 'position': 'Anywhere', 'classification': 'Artefact', 'note': 'Convertion of glycosylated asparagine residues upon deglycosylation with PNGase F in H2O'}, {'hidden': True, 'spec_group': 3, 'site': 'F', 'position': 'Protein N-term', 'classification': 'Post-translational'}], 'alt_names': ['phenyllactyl from N-term Phe', 'Citrullination'], 'refs': [{'text': '6838602', 'source': 'PubMed PMID', 'url': None}, {'text': 'AA0214', 'source': 'RESID', 'url': None}, {'text': 'IonSource tutorial', 'source': 'Misc. URL', 'url': 'http://www.ionsource.com/Card/Deamidation/deamidation.htm'}, {'text': 'CITR', 'source': 'FindMod', 'url': None}, {'text': 'AA0128', 'source': 'RESID', 'url': None}, {'text': 'FLAC', 'source': 'FindMod', 'url': None}, {'text': '15700232', 'source': 'PubMed PMID', 'url': None}, {'text': 'DEAM', 'source': 'FindMod', 'url': None}]}
    2023-10-27 15:44:48.049 | DEBUG    | ms2ml.proforma_utils:mod_id_mass:117 - Resolving mod_id: 1
    2023-10-27 15:44:48.049 | DEBUG    | ms2ml.proforma_utils:mod_id_mass:119 - Resolved to {'date_time_modified': datetime.datetime(2017, 11, 8, 16, 8, 56), 'date_time_posted': datetime.datetime(2002, 8, 19, 19, 17, 11), 'avge_mass': 42.0367, 'mono_mass': 42.010565, 'composition': Composition({'H': 2, 'C': 2, 'O': 1}), 'record_id': 1, 'approved': True, 'title': 'Acetyl', 'full_name': 'Acetylation', 'username_of_poster': 'unimod', 'group_of_poster': 'admin', 'specificity': [{'hidden': True, 'spec_group': 6, 'site': 'T', 'position': 'Anywhere', 'classification': 'Post-translational'}, {'hidden': False, 'spec_group': 5, 'site': 'N-term', 'position': 'Protein N-term', 'classification': 'Post-translational'}, {'hidden': True, 'spec_group': 4, 'site': 'S', 'position': 'Anywhere', 'classification': 'Post-translational'}, {'hidden': True, 'spec_group': 3, 'site': 'C', 'position': 'Anywhere', 'classification': 'Post-translational'}, {'hidden': False, 'spec_group': 2, 'site': 'N-term', 'position': 'Any N-term', 'classification': 'Multiple', 'note': 'GIST acetyl light'}, {'hidden': False, 'spec_group': 1, 'site': 'K', 'position': 'Anywhere', 'classification': 'Multiple', 'note': 'PT and GIST acetyl light'}, {'hidden': True, 'spec_group': 7, 'site': 'Y', 'position': 'Anywhere', 'classification': 'Chemical derivative', 'note': 'O-acetyl'}, {'hidden': True, 'spec_group': 8, 'site': 'H', 'position': 'Anywhere', 'classification': 'Chemical derivative'}, {'hidden': True, 'spec_group': 9, 'site': 'R', 'position': 'Anywhere', 'classification': 'Artefact', 'note': 'glyoxal-derived hydroimidazolone'}], 'refs': [{'text': 'AA0048', 'source': 'RESID', 'url': None}, {'text': 'AA0049', 'source': 'RESID', 'url': None}, {'text': 'AA0041', 'source': 'RESID', 'url': None}, {'text': 'AA0052', 'source': 'RESID', 'url': None}, {'text': 'AA0364', 'source': 'RESID', 'url': None}, {'text': 'AA0056', 'source': 'RESID', 'url': None}, {'text': 'AA0046', 'source': 'RESID', 'url': None}, {'text': 'AA0051', 'source': 'RESID', 'url': None}, {'text': 'AA0045', 'source': 'RESID', 'url': None}, {'text': 'AA0354', 'source': 'RESID', 'url': None}, {'text': 'AA0044', 'source': 'RESID', 'url': None}, {'text': 'AA0043', 'source': 'RESID', 'url': None}, {'text': '11999733', 'source': 'PubMed PMID', 'url': None}, {'text': 'Chemical Reagents for Protein Modification 3rd edition, pp 215-221, Roger L. Lundblad, CRC Press, New York, N.Y., 2005', 'source': 'Book', 'url': None}, {'text': 'IonSource acetylation tutorial', 'source': 'Misc. URL', 'url': 'http://www.ionsource.com/Card/acetylation/acetylation.htm'}, {'text': 'AA0055', 'source': 'RESID', 'url': None}, {'text': '14730666', 'source': 'PubMed PMID', 'url': None}, {'text': '15350136', 'source': 'PubMed PMID', 'url': None}, {'text': 'AA0047', 'source': 'RESID', 'url': None}, {'text': '12175151', 'source': 'PubMed PMID', 'url': None}, {'text': '11857757', 'source': 'PubMed PMID', 'url': None}, {'text': 'AA0042', 'source': 'RESID', 'url': None}, {'text': 'AA0050', 'source': 'RESID', 'url': None}, {'text': 'AA0053', 'source': 'RESID', 'url': None}, {'text': 'AA0054', 'source': 'RESID', 'url': None}, {'text': 'ACET', 'source': 'FindMod', 'url': None}, {'text': 'PNAS 2006 103: 18574-18579', 'source': 'Journal', 'url': 'http://dx.doi.org/10.1073/pnas.0608995103'}]}
:::

::: {.cell-output .cell-output-display execution_count="9"}
    array([200.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
             0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
             0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
             0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
             0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
             0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
             0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
             0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
             0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
             0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
             0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
          dtype=float32)
:::
:::

::: {.cell execution_count="10"}
``` {.python .cell-code}
print("First 5: ", spectrum.fragment_labels[:5])
print("Last 5:", spectrum.fragment_labels[-5:])
print("Total Number:", len(spectrum.fragment_labels))
```

::: {.cell-output .cell-output-stdout}
    First 5:  ['y1^1', 'y1^2', 'y2^1', 'y2^2', 'y3^1']
    Last 5: ['b28^2', 'b29^1', 'b29^2', 'b30^1', 'b30^2']
    Total Number: 120
:::
:::

#### Modifying the encoding schema

::: {.cell execution_count="11"}
``` {.python .cell-code}
random_config = Config(ion_series = 'zc', ion_charges = (1,))
spectrum.config = random_config
print(spectrum.encode_fragments())
print("First 5: ", spectrum.fragment_labels[:5])
print("Last 5:", spectrum.fragment_labels[-5:])
print("Total Number:", len(spectrum.fragment_labels))
```

::: {.cell-output .cell-output-stdout}
    [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
     0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
     0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
    First 5:  ['z1^1', 'z2^1', 'z3^1', 'z4^1', 'z5^1']
    Last 5: ['c26^1', 'c27^1', 'c28^1', 'c29^1', 'c30^1']
    Total Number: 60
:::
:::


docs/ms2ml/configuration.md


# Reference

::: ms2ml.config.Config
    handler: python


ms2ml/config.py
"""Provides a place to define the project's configuration.

Defines the constants that are used in the rest of the project.

Such as the masses of aminoacids, supported modifications, length of the encodings,
maximum length supported, labels and order of the encoded ions ...
"""

from __future__ import annotations

import string
import sys
import warnings
from dataclasses import asdict, dataclass, field

if sys.version_info <= (3, 11):
    import tomli as tomllib
else:
    import tomllib

import numpy as np
import tomli_w

from .annotation_classes import AnnotatedIon
from .constants import C_TERMINUS, N_TERMINUS, STD_AA_MASS
from .proforma_utils import MemoizedUnimodResolver
from .type_defs import MassError, ModModes
from .utils.class_utils import lazy

# TODO cosnsider wether we want a more strict enforcement
# of the requirement for configurations to be defined
# and passed


def _default_mod_aliases():
    out = {
        "Phospho": "[UNIMOD:21]",
        "TMT6plex": "[UNIMOD:737]",
        "TMT10plex": "[UNIMOD:737]",
        "Deamidation": "[UNIMOD:7]",
        "Acetyl": "[UNIMOD:1]",
        "Oxidation": "[UNIMOD:35]",
        "Hydroxilation": "[UNIMOD:35]",
        "Trimethyl": "[UNIMOD:37]",
    }
    return out


def _default_mod_order():
    encoding_mod_order = tuple(
        [
            None,
            "[UNIMOD:4]",
            "[UNIMOD:21]",
            "[UNIMOD:35]",
            "[UNIMOD:7]",
            "[UNIMOD:1]",
            "__unknown1__",
        ]
    )
    return encoding_mod_order


def _default_var_mods():
    out = {
        "[UNIMOD:35]": [
            "M",
        ],
        "[UNIMOD:21]": [
            "S",
            "T",
            "Y",
        ],
    }
    return out


@dataclass
class Config:
    """General class to set and store the configuration of the project.

    Ideally every project will make one AND ONLY ONE of these.
    The provided defaults are meant to be reasonable for most projects
    but can be changed as needed.

    All annotation and endoding functionality should require one of this objects
    to work.

    Parameters:
        g_tolerances:
            A tuple of floats,
            where each float is the tolerance of that corresponding ms level.
            For instance (10, 20) means that the tolerance for ms1 is 10, 20 for ms2.
        g_tolerance_units:
            A tuple of strings, that denote what tolerance unit to use for each ms
            level. For instance ("ppm", "Da") means that the tolerance for ms1 is
            in ppm, and for ms2 in Da.
        g_isotopes:
        peptide_length_range:
            A tuple of ints, where the first int is the minimum length of peptides.
        precursor_charges:
            A tuple of ints, where each int is a possible precursor charge.
        fragment_positions:
            A tuple of ints, where each int is a possible fragment position.
        ion_series:
            A string of characters, where each character is a possible ion series.
            An example,a dn the default is ('by')
        ion_charges:
            A tuple of ints, where each int is a possible ion charge.
        ion_neutral_losses:
        ion_encoding_nesting:
        ion_naming_convention:
        mod_ambiguity_threshold:
        mod_fixed_mods:

    Examples:
        >>> config = Config()
        >>> config
        Config(g_tolerances=(50, 50), ...)
        >>> config.fragment_labels
        ['y1^1', 'y1^2', ... 'b30^2']
    """

    # General Configs
    # = each tuple object is meant to define an MS level
    # = For instance, in tolerances, the first tolerance is for MS1
    # = and the second for MS2
    g_tolerances: tuple[float, ...] = (50, 50)

    g_tolerance_units: tuple[MassError, ...] = ("ppm", "ppm")

    # = Number of isotopes to check for each ion, 0 means that only
    # = the monoisotopic peak us used
    g_isotopes: tuple[int, ...] = (0, 0)

    # Peptide Configs
    peptide_length_range: tuple[int, int] = (5, 30)
    peptide_mz_range: tuple[float, float] = (200, 20000)

    # Precursor Configs
    precursor_charges: tuple[int, ...] = (1, 2, 3, 4, 5, 6)

    # Fragment Configs
    # TODO consider wether to unify this the same way the general configs
    # = use positions within a tuple to denote MS levels
    fragment_positions: tuple[int, ...] = tuple(range(1, 31))

    # Ion config
    ion_series: str = "yb"
    ion_charges: tuple[int, ...] = (1, 2)
    # = Neutral Losss Configs
    ion_neutral_losses: tuple[str, ...] = ()

    ion_encoding_nesting: tuple[str, ...] = (
        "ion_charges",
        "fragment_positions",
        "ion_series",
    )
    ion_naming_convention: str = "{ion_series}{fragment_positions}^{ion_charges}"

    # Modifications
    mod_mode: ModModes = "unimod"
    mod_ambiguity_threshold: float = 0.99
    mod_fixed_mods: tuple[str] = ("[UNIMOD:4]@C",)
    mod_variable_mods: dict[str, tuple[str]] = field(default_factory=_default_var_mods)

    # Encoding Configs
    encoding_aa_order: tuple[str] = tuple(
        ["n_term", *list(string.ascii_uppercase), "c_term", "__missing__"]
    )

    encoding_mod_order: tuple[str | None, ...] = field(
        default_factory=_default_mod_order
    )
    encoding_mod_alias: dict[str, str] = field(default_factory=_default_mod_aliases)

    encoding_spec_bin_start: float = field(repr=False, default=0.0)
    encoding_spec_bin_end: float = field(repr=False, default=2000.0)
    encoding_spec_bin_binsize: float | None = field(repr=False, default=0.1)
    encoding_spec_bin_n_bins: int | None = field(repr=False, default=None)
    encoding_spec_bin_relative: bool = field(repr=False, default=False)
    encoding_spec_bin_offset: float = field(repr=False, default=0.0)

    @lazy
    def fragment_labels(self) -> list[str]:
        """
        Returns a list of the labels that are used to encode the fragments.

        Examples:
            >>> config = Config()
            >>> config.fragment_labels
            ['y1^1', 'y1^2', ... 'b30^2']
        """
        labels: list[str] = []
        for ion_field in self.ion_encoding_nesting:
            _labels = labels
            labels = []
            for elem in getattr(self, ion_field):
                mapper = _PermissiveMapper(**{ion_field: str(elem)})
                if len(_labels) == 0:
                    x = self.ion_naming_convention.format_map(mapper)
                    labels.append(x)
                else:
                    for x in _labels:
                        x = x.format_map(mapper)
                        labels.append(x)
        return labels

    @property
    def num_fragment_embeddings(self) -> int:
        return len(self.fragment_labels)

    def ion_labeller(self, ion: AnnotatedIon) -> str:
        """Labels an ion

        Provided an ion, returns the label for that ion.

        Examples:
            >>> ion = AnnotatedIon(mass=123.2, charge=2, position=3, ion_series="z")
            >>> config = Config()
            >>> config.ion_labeller(ion)
            'z3^2'
        """
        return self.ion_naming_convention.format_map(ion.asdict())

    @lazy
    def aa_masses(self):
        masses = STD_AA_MASS.copy()
        masses["n_term"] = N_TERMINUS
        masses["c_term"] = C_TERMINUS
        aa_mass = [masses.get(aa, 0) for aa in self.encoding_aa_order]
        return np.array(aa_mass)

    @lazy
    def mod_masses(self):
        mod_mass = []
        for mod in self.encoding_mod_order:
            if mod is None:
                mod_mass.append(0.0)

            elif "[UNIMOD:" in mod:
                mod_id = int(mod.split(":")[1].split("]")[0])
                mod_mass.append(MemoizedUnimodResolver.mod_id_mass(mod_id))

            elif "unknown" in mod:
                mod_mass.append(0.0)

            else:
                raise ValueError(f"Unknown mod {mod}")

        return np.array(mod_mass)

    @lazy
    def encoding_mod_order_mapping(self) -> dict[str | None, int]:
        return {mod: i for i, mod in enumerate(self.encoding_mod_order)}

    @lazy
    def encoding_aa_order_mapping(self) -> dict[str | None, int]:
        return {aa: i for i, aa in enumerate(self.encoding_aa_order)}

    def _resolve_mod_list(self, x):
        if isinstance(x, list):
            return [self._resolve_mod_list(y) for y in x]

        if x is None:
            return None

        if self.mod_mode == "unimod":
            return self._resolve_mod_list_unimod(x)
        elif self.mod_mode == "delta_mass":
            return self._resolve_mod_list_delta_mass(x)
        else:
            raise NotImplementedError(
                f"Mod mode {self.mod_mode} not implemented. Trying to resolve {x}"
            )

    def _resolve_mod_list_delta_mass(self, x):
        val = x.value
        if isinstance(val, float):
            return val

        if ":" in val and "Obs:" not in val:
            raise ValueError(f"Delta mass mods should not contain a colon. Got {val}")
        if "Obs:" in val:
            val = val.replace("Obs:", "")

        if not (val.startswith("+") or val.startswith("-")):
            val = "+" + val

        return val

    def _resolve_mod_list_unimod(self, x):
        """Resolves the names in a list of modifications to unimod Ids."""

        if not hasattr(self, "__mod_resol_cache"):
            setattr(self, "__mod_resol_cache", {None: None})

        cache = getattr(self, "__mod_resol_cache")

        if x.value in cache:
            return cache[x.value]

        # TODO consider moving this logic to the mod resolver class
        def _internal(y):
            # Mass modififications (only deinfed by mass, such as open mods)
            # or underfined aliases .... do not have a name

            # Write a better error message if the mod is not found
            if y.value in self.encoding_mod_alias:
                modname = y.value
            else:  # noqa
                if not hasattr(y, "name"):
                    modname = str(y)
                else:
                    modname = y.name

            if modname in self.encoding_mod_alias:
                solved_name = self.encoding_mod_alias[modname]
            else:
                solved_name = MemoizedUnimodResolver.resolve(modname)["id"]
                solved_name = f"[UNIMOD:{solved_name!s}]"

            return solved_name

        cache[x.value] = _internal(x)
        return cache[x.value]

    def asdict(self):
        """Returns a dictionary representation of the config."""
        return asdict(self)

    """
    # def validate(self):
    # test_all_modification_aliases_map():
    for v in constants.MOD_PEPTIDE_ALIASES.values():
        if v == "":
            continue
        constants.MODIFICATION[v]

    # Validate ion naming positions ...
    # Test all variable mods are in the mod order
    # Test all mod aliases in mod order
    # Test all aas in mod aliases in aa_order
    """
    #     raise NotImplementedError

    def to_toml(self, path: str):
        """Writes the config to a toml file."""

        out = {}
        for k, v in self.asdict().items():
            if isinstance(v, tuple):
                v = [x if x is not None else "__NONE__" for x in v]
            elif v is None:
                v = "__NONE__"
            out[k] = v
        with open(path, "wb") as f:
            tomli_w.dump(out, f)

    @staticmethod
    def from_toml(path: str):
        """Loads a config from a toml file."""
        with open(path, "rb") as f:
            config = tomllib.load(f)

        out_config = {}
        for k, v in config.items():
            if isinstance(v, list):
                v = tuple(x if x != "__NONE__" else None for x in v)
            elif v == "__NONE__":
                v = None
            out_config[k] = v

        return Config(**out_config)

    def from_comet(self, path: str, *args, **kwargs):
        """Loads a config from a comet params file."""
        raise NotImplementedError


class _PermissiveMapper(dict):
    """This class is a helper to allow the bypass of missing keys in the generation of.

    the ion labels.
    """

    def __missing__(self, key):
        return "{" + key + "}"


def get_default_config() -> Config:
    warnings.warn("Using default config. Consider creating your own.", UserWarning)
    return Config()


class ConfigNotSetError(Exception):
    """Raised when a config is not set but is required."""
